sql/rowexec: TestDrainingProcessorSwallowsUncertaintyError failed

sql/rowexec.TestDrainingProcessorSwallowsUncertaintyError [failed](https://tanzanite.cluster.engflow.com/invocations/default/ed4c1d3d-62fb-4e38-9fa4-6341741d7fc0?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL3Jvd2V4ZWM6cm93ZXhlY190ZXN0) on master @ [f0116ea373a2b87155e7f0264df4f783ce177360](https://github.com/cockroachdb/cockroach/commits/f0116ea373a2b87155e7f0264df4f783ce177360):

Fatal error:

```
panic: test timed out after 41m37s
running tests:
	TestDrainingProcessorSwallowsUncertaintyError (40m57s)

```
Stack: 

```
goroutine 24845 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x265
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x45
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* github.com/cockroachdb/pebble/record.NewLogWriter.func2()
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:562 +0x92
* created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 24212
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:561 +0x8e5
* 
* goroutine 23505 [sync.Cond.Wait, 2 minutes]:
* sync.runtime_notifyListWait(0xc0097000a8, 0x4e)
* 	GOROOT/src/runtime/sema.go:569 +0x159
* sync.(*Cond).Wait(0xc009700098)
* 	GOROOT/src/sync/cond.go:70 +0x75
* github.com/cockroachdb/pebble/record.(*flusherCond).Wait(...)
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:369
* github.com/cockroachdb/pebble/record.(*LogWriter).flushLoop(0xc009700000, {0xc009565778?, 0x7b9d240?})
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:643 +0xb79
* runtime/pprof.Do({0xb2a1218, 0xfd662c0}, {{0xc000d7bc60?, 0xc005bbea20?, 0xc003cf63f0?}}, 0xc001c6dfc0)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
* github.com/cockroachdb/pebble/record.NewLogWriter.func2()
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:562 +0x92
* created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 20656
* 	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:561 +0x8e5
* 
* goroutine 24084 [select, 2 minutes]:
* database/sql.(*DB).connectionOpener(0xc0055a60d0, {0xb2a19a0, 0xc0084b6190})
* 	GOROOT/src/database/sql/sql.go:1246 +0xed
* created by database/sql.OpenDB in goroutine 20401
* 	GOROOT/src/database/sql/sql.go:824 +0x2d4
* 
* goroutine 24091 [select, 2 minutes]:
* database/sql.(*DB).connectionOpener(0xc0055a6270, {0xb2a19a0, 0xc0084b6280})
* 	GOROOT/src/database/sql/sql.go:1246 +0xed
* created by database/sql.OpenDB in goroutine 20401
* 	GOROOT/src/database/sql/sql.go:824 +0x2d4
* 
* goroutine 24098 [select, 2 minutes]:
* database/sql.(*DB).connectionOpener(0xc0055a64e0, {0xb2a19a0, 0xc0084b6370})
* 	GOROOT/src/database/sql/sql.go:1246 +0xed
* created by database/sql.OpenDB in goroutine 20401
* 	GOROOT/src/database/sql/sql.go:824 +0x2d4
* 
* goroutine 24815 [sleep]:
* time.Sleep(0x4c4b40)
* 	GOROOT/src/runtime/time.go:195 +0x115
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Quiesce(0xc004d40820, {0xb2a19d8, 0xc003cf6870})
* 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:604 +0x31e
* github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers.func1(0x0, 0xc004d40820)
* 	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:156 +0x17f
* created by github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).stopServers in goroutine 20401
* 	github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:152 +0x1cf
* 
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDrainingProcessorSwallowsUncertaintyError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-36849

sql/rowexec.TestDrainingProcessorSwallowsUncertaintyError [failed](https://tanzanite.cluster.engflow.com/invocations/default/d51b4032-b124-4314-bae9-dec40227dc97?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL3Jvd2V4ZWM6cm93ZXhlY190ZXN0) on master @ [0dce29441187eda187aad472543997b06c2b01ea](https://github.com/cockroachdb/cockroach/commits/0dce29441187eda187aad472543997b06c2b01ea):

Fatal error:

```
panic: test timed out after 41m37s
running tests:
	TestDrainingProcessorSwallowsUncertaintyError (40m57s)
	TestDrainingProcessorSwallowsUncertaintyError/vectorize=false (40m50s)
	TestDrainingProcessorSwallowsUncertaintyError/vectorize=false/dummy=true (40m50s)

```
Stack: 

```
goroutine 966426 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x265
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x45
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestDrainingProcessorSwallowsUncertaintyError
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestDrainingProcessorSwallowsUncertaintyError2037115885
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)

pkg/sql/rowexec/processors_test.go:521: (TestDrainingProcessorSwallowsUncertaintyError)
	NOTICE: .GetFirstStoreID() called via implicit interface StorageLayerInterface;
HINT: consider using .StorageLayer().GetFirstStoreID() instead.

pkg/sql/rowexec/processors_test.go:522: (TestDrainingProcessorSwallowsUncertaintyError)
	NOTICE: .GetFirstStoreID() called via implicit interface StorageLayerInterface;
HINT: consider using .StorageLayer().GetFirstStoreID() instead.

pkg/sql/rowexec/processors_test.go:528: (TestDrainingProcessorSwallowsUncertaintyError)
	NOTICE: .AdvSQLAddr() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().AdvSQLAddr() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
=== RUN   TestDrainingProcessorSwallowsUncertaintyError/vectorize=false/dummy=false
=== RUN   TestDrainingProcessorSwallowsUncertaintyError/vectorize=false/dummy=true
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDrainingProcessorSwallowsUncertaintyError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/sql/rowexec/processors_test.go
+++ b/pkg/sql/rowexec/processors_test.go
@@ -447,14 +447,14 @@ func TestDrainingProcessorSwallowsUncertaintyError(t *testing.T) {
 	// entirely easy to cause given the current implementation details.
 
 	var (
-		// trapRead is set, atomically, once the test wants to block a read on the
-		// first node.
-		trapRead    int64
-		blockedRead struct {
+		// We will want to block exactly one read to the first node.
+		blockOneRead atomic.Bool
+		blockedRead  struct {
 			syncutil.Mutex
 			unblockCond   *sync.Cond
 			shouldUnblock bool
 		}
+		endKeyToBlock string
 	)
 
 	blockedRead.unblockCond = sync.NewCond(&blockedRead.Mutex)
@@ -470,7 +470,7 @@ func TestDrainingProcessorSwallowsUncertaintyError(t *testing.T) {
 					Knobs: base.TestingKnobs{
 						Store: &kvserver.StoreTestingKnobs{
 							TestingRequestFilter: func(_ context.Context, ba *kvpb.BatchRequest) *kvpb.Error {
-								if atomic.LoadInt64(&trapRead) == 0 {
+								if !blockOneRead.Load() {
 									return nil
 								}
 								// We're going to trap a read for the rows [1,5].
@@ -478,23 +478,28 @@ func TestDrainingProcessorSwallowsUncertaintyError(t *testing.T) {
 								if !ok {
 									return nil
 								}
-								key := req.(*kvpb.ScanRequest).Key.String()
 								endKey := req.(*kvpb.ScanRequest).EndKey.String()
-								if strings.Contains(key, "/1") && strings.Contains(endKey, "/6") {
-									blockedRead.Lock()
-									for !blockedRead.shouldUnblock {
-										blockedRead.unblockCond.Wait()
-									}
-									blockedRead.Unlock()
-									return kvpb.NewError(
-										kvpb.NewReadWithinUncertaintyIntervalError(
-											ba.Timestamp,           /* readTs */
-											hlc.ClockTimestamp{},   /* localUncertaintyLimit */
-											ba.Txn,                 /* txn */
-											ba.Timestamp.Add(1, 0), /* valueTS */
-											hlc.ClockTimestamp{} /* localTS */))
+								if !strings.Contains(endKey, endKeyToBlock) {
+									// Either a request for a different table or
+									// for a different part of the target table.
+									return nil
 								}
-								return nil
+								// Since we're blocking this read, the caller
+								// must set it again if it wants to block
+								// another read.
+								blockOneRead.Store(false)
+								blockedRead.Lock()
+								for !blockedRead.shouldUnblock {
+									blockedRead.unblockCond.Wait()
+								}
+								blockedRead.Unlock()
+								return kvpb.NewError(
+									kvpb.NewReadWithinUncertaintyIntervalError(
+										ba.Timestamp,           /* readTs */
+										hlc.ClockTimestamp{},   /* localUncertaintyLimit */
+										ba.Txn,                 /* txn */
+										ba.Timestamp.Add(1, 0), /* valueTS */
+										hlc.ClockTimestamp{} /* localTS */))
 							},
 						},
 					},
@@ -510,6 +515,15 @@ func TestDrainingProcessorSwallowsUncertaintyError(t *testing.T) {
 		"x INT PRIMARY KEY",
 		10, /* numRows */
 		sqlutils.ToRowFn(sqlutils.RowIdxFn))
+
+	row := origDB0.QueryRow("SELECT 't'::regclass::oid")
+	var tableID int
+	require.NoError(t, row.Scan(&tableID))
+	// Request that we want to block is of the form
+	//   Scan /Table/105/1{-/6}
+	// so it'll have the end key like
+	//   /Table/105/1/6.
+	endKeyToBlock = fmt.Sprintf("/Table/%d/1/6", tableID)
 
 	// Split the table and move half of the rows to the 2nd node. We'll block the
 	// read on the first node, and so the rows we're going to be expecting are the
@@ -528,8 +542,6 @@ func TestDrainingProcessorSwallowsUncertaintyError(t *testing.T) {
 	defaultConn, cleanup := getPGXConnAndCleanupFunc(ctx, t, tc.Server(0).AdvSQLAddr())
 	defer cleanup()
 
-	atomic.StoreInt64(&trapRead, 1)
-
 	// Run with the vectorize off and on.
 	testutils.RunTrueAndFalse(t, "vectorize", func(t *testing.T, vectorize bool) {
 		// We're going to run the test twice in each vectorize configuration. Once
@@ -537,6 +549,7 @@ func TestDrainingProcessorSwallowsUncertaintyError(t *testing.T) {
 		// by increasing the limit from 5 to 6 and checking that we get the injected
 		// error in that case.
 		testutils.RunTrueAndFalse(t, "dummy", func(t *testing.T, dummy bool) {
+			blockOneRead.Store(true)
 			// Reset the blocking condition.
 			blockedRead.Lock()
 			blockedRead.shouldUnblock = false

